tools/state-viewer/src/commands.rs

```diff
@@ -649,3 +659,168 @@ pub fn format_hash(h: CryptoHash) -> String {
 pub fn chunk_mask_to_str(mask: &[bool]) -> String {
     mask.iter().map(|f| if *f { '.' } else { 'X' }).collect()
 }
+
+// like ChainStoreUpdate::get_incoming_receipts_for_shard(), but for the case when we don't
+// know of a block containing the target chunk
+fn get_incoming_receipts(
+    chain_store: &mut ChainStore,
+    chunk_hash: &ChunkHash,
+    shard_id: u64,
+    target_height: u64,
+    prev_hash: &CryptoHash,
+    prev_height_included: u64,
+) -> anyhow::Result<Vec<Receipt>> {
+    let mut receipt_proofs = vec![];
+
+    let chunk_hashes = chain_store.get_all_chunk_hashes_by_height(target_height)?;
+    if !chunk_hashes.contains(chunk_hash) {
+        return Err(anyhow!(
+            "given chunk hash is not listed in ColChunkHashesByHeight[{}]",
+            target_height
+        ));
+    }
+
+    let chunks =
+        chunk_hashes.iter().map(|h| chain_store.get_chunk(h).unwrap().clone()).collect::<Vec<_>>();
+
+    for chunk in chunks {
+        let partial_encoded_chunk = chain_store.get_partial_chunk(&chunk.chunk_hash()).unwrap();
+        for receipt in partial_encoded_chunk.receipts().iter() {
+            let ReceiptProof(_, shard_proof) = receipt;
+            if shard_proof.to_shard_id == shard_id {
+                receipt_proofs.push(receipt.clone());
+            }
+        }
+    }
+    let mut responses = vec![ReceiptProofResponse(CryptoHash::default(), receipt_proofs)];
+    responses.extend_from_slice(&chain_store.store_update().get_incoming_receipts_for_shard(
+        shard_id,
+        *prev_hash,
+        prev_height_included,
+    )?);
+    Ok(collect_receipts_from_response(&responses))
+}
+
+fn print_apply_chunk_result(
+    result: ApplyTransactionResult,
+    gas_limit: Gas,
+    tx_hashes: Option<Vec<CryptoHash>>,
+    receipt_hashes: Option<Vec<CryptoHash>>,
+) {
+    if tx_hashes.is_some() || receipt_hashes.is_some() {
+        let mut hashes = HashSet::new();
+        if let Some(tx_hashes) = tx_hashes {
+            hashes.extend(tx_hashes);
+        }
+        if let Some(receipt_hashes) = receipt_hashes {
+            hashes.extend(receipt_hashes);
+        }
+
+        println!("outcomes:");
+        for outcome in result.outcomes.iter() {
+            if hashes.contains(&outcome.id) {
+                println!("{:?}", outcome);
+            }
+        }
+    }
+    println!("resulting chunk extra:\n{:?}", resulting_chunk_extra(result, gas_limit));
+}
+
+fn check_hashes_exist<T, F: Fn(&T) -> CryptoHash>(
+    hashes: &Option<Vec<CryptoHash>>,
+    items: &[T],
+    item_hash: F,
+) -> Result<(), CryptoHash> {
+    match hashes {
+        Some(hashes) => {
+            let hashes_seen = items.iter().map(item_hash).collect::<HashSet<_>>();
+            for hash in hashes.iter() {
+                if !hashes_seen.contains(hash) {
+                    return Err(*hash);
+                }
+            }
+            Ok(())
+        }
+        None => Ok(()),
+    }
+}
+
+pub(crate) fn apply_chunk(
```

Comment:
The new command requires quite a lot of logic, please consider moving it to a separate file.
And if you move it to a separate file, would it be possible to add a simple unit test?

Reply:
moved and added test